packages/graph/hash_graph/lib/graph/src/store/postgres/ontology/entity_type/mod.rs

```diff
@@ -1,15 +1,120 @@
 mod resolve;
 
+use std::{
+    collections::{hash_map::Entry, HashMap},
+    future::Future,
+    pin::Pin,
+};
+
 use async_trait::async_trait;
 use error_stack::{IntoReport, Result, ResultExt};
+use futures::{stream, FutureExt, StreamExt, TryStreamExt};
 use tokio_postgres::GenericClient;
-use type_system::EntityType;
+use type_system::{uri::VersionedUri, EntityType};
 
 use crate::{
-    ontology::{AccountId, PersistedOntologyIdentifier},
-    store::{AsClient, EntityTypeStore, InsertionError, PostgresStore, UpdateError},
+    ontology::{
+        AccountId, EntityTypeTree, PersistedDataType, PersistedEntityType, PersistedLinkType,
+        PersistedOntologyIdentifier, PersistedPropertyType,
+    },
+    store::{
+        crud::Read,
+        postgres::{context::PostgresContext, PersistedOntologyType},
+        query::Expression,
+        AsClient, EntityTypeStore, InsertionError, PostgresStore, QueryError, UpdateError,
+    },
 };
 
+impl<C: AsClient> PostgresStore<C> {
+    /// Internal method to read a [`PersistedEntityType`] into a [`HashMap`].
+    ///
+    /// This is used to recursively resolve a type, so the result can be reused.
+    #[expect(
+        clippy::too_many_arguments,
+        reason = "Output is created in place and needs to be reused later"
+    )]
+    pub(crate) fn get_entity_type_as_dependency<'a>(
+        &'a self,
+        entity_type_uri: VersionedUri,
+        data_type_references: &'a mut HashMap<VersionedUri, PersistedDataType>,
+        property_type_references: &'a mut HashMap<VersionedUri, PersistedPropertyType>,
+        link_type_references: &'a mut HashMap<VersionedUri, PersistedLinkType>,
+        entity_type_references: &'a mut HashMap<VersionedUri, PersistedEntityType>,
+        data_type_resolve_depth: u8,
+        property_type_resolve_depth: u8,
+        link_type_query_depth: u8,
+        entity_type_query_depth: u8,
+    ) -> Pin<Box<dyn Future<Output = Result<(), QueryError>> + Send + 'a>> {
+        async move {
+            // TODO: Avoid cloning of URI
+            //   see https://stackoverflow.com/questions/51542024
+            if let Entry::Vacant(entry) = entity_type_references.entry(entity_type_uri) {
+                let entity_type = PersistedEntityType::from_record(
+                    self.read_versioned_ontology_type::<EntityType>(entry.key())
+                        .await?,
+                );
+                let entity_type = entry.insert(entity_type);
+
+                if let Some(new_depth) = property_type_resolve_depth.checked_sub(1) {
```

Comment:
100% not a request for changes:

I wonder if `checked_sub` is the right pattern here. I appreciate it lets us do a clean one-liner but I could see worlds where this actually leads to funky bugs.

`checked_sub` is also a method on `i8`, so if for some reason we were to swap the integer type used for the depth parameters, I think this method would then fail weirdly as we'd call the thing with `-1`.

Generally speaking I think it's clearer if we explicitly say what we're checking, that is, the number is at least 0. Rather than some obscure thing about overflows.